### PR TITLE
Handle eventId url parameter

### PR DESCRIPTION
**Pull request by Abyss777**
Also fit map view to all markers not only first route.

The logic is not obvious because of asynchronous. I'll try to explain.

- `Application` add function to remove parameter from url and global variable.
- `ReportController.init` check if url has `eventId` parameter and initiate event load process.
Remove `eventId` from url
- `MapMarkerController.loadReport` removed map view fit
- `MapMarkerController.addReportMarkers` calculate `ol.Extent` to fit map view.
Fire 'selectreport' event in case we come here with `eventId` parameter

After this changes if we pass parameter, event will be loaded, position will be loaded and showed on map.

I was thinking, if event does not have linked position, user won't understand what's happen.

- `ReportController.loadEvents` open report panel and highlight event in case we do not have linked position.

### CODE REVIEW

**Comment by tananaev** [comment]
Is this intended for email links to show events?

**Comment by Abyss777** [comment]
Yes. It is the main reason.

**Comment by tananaev** [comment]
The logic is very weird. I don't like that we use global variable and then manipulate URL from report controller. There must be a better way.

**Comment by Abyss777** [comment]
We can do it without manipulating url. It is just for user comfort, not load event again on page refresh.

**Comment by tananaev** [comment]
I'm not against URL manipulation, but it should be in some more appropriate place. And we should pass parameters around instead of using global variables as much as possible.

**Comment by Abyss777** [comment]
I agree with you but everything is asynchronous. Callbacks is everywhere.
Another problem is that everything is ready in different time. Some events are missed by some components on the load stage.

But I will try again.

**Comment by tananaev** [comment]
Let me know if you need any help from me.

**Comment by Abyss777** [comment]
is Root.js more appropriate place to handle url parameters and url manipulation?

**Comment by tananaev** [comment]
Yes, I think so. Then you can pass the parameter down to map and/or report
somehow.

**Comment by Abyss777** [comment]
Implemented it in other way.

fire `showsinlgleevent` in Root.js
catch it in `ReportController` and initiate load process
in callback select row in report and expand if needed

I need to temporary store variable in `ReportController` to show position in State panel

File: web/app/view/MapMarkerController.js [hunk 1]
```diff
@@ -202,37 +202,37 @@ Ext.define('Traccar.view.MapMarkerController', {
...
-        var i, position, point, geometry, marker, style;
+        var i, position, point, geometry, marker, style, minx, miny, maxx, maxy;
         this.clearReport();
         for (i = 0; i < data.length; i++) {
             position = data[i];
             point = ol.proj.fromLonLat([
                 position.get('longitude'),
                 position.get('latitude')
             ]);
+            minx = minx ? Math.min(point[0], minx) : point[0];
```
Comment (tananaev) [inline web/app/view/MapMarkerController.js]: Why do we need this logic?

File: web/app/view/ReportController.js [hunk 2]
```diff
@@ -208,15 +210,45 @@ Ext.define('Traccar.view.ReportController', {
                 params: {
                     id: positionIds
                 },
+                scope: this,
                 callback: function (records, operation, success) {
                     if (success) {
                         Ext.getStore('ReportRoute').add(records);
+                        if (this.singleEvent && records.length > 0) {
```
Comment (tananaev) [inline web/app/view/ReportController.js]: I'm think maybe it's better to do it all reports anyway? Then we don't need `singleEvent` variable and everything is going to be simpler. What do you think?

File: web/app/controller/Root.js [hunk 3]
```diff
@@ -94,6 +94,11 @@ Ext.define('Traccar.controller.Root', {
         } else {
             Ext.create('widget.main');
         }
+        eventId = Ext.Object.fromQueryString(window.location.search).eventId;
+        if (eventId) {
+            this.fireEvent('showsingleevent', eventId);
+            //this.removeUrlParameter('eventId');
```
Comment (tananaev) [inline web/app/controller/Root.js]: Why is this commented out?

File: web/app/view/MapMarkerController.js (hunk 1 above)
Comment (Abyss777) [inline web/app/view/MapMarkerController.js]: The only reason is `Math.min(point[0], undefined)` gives `NaN`.
I will redo it with less compares.

File: web/app/view/ReportController.js (hunk 2 above)
Comment (Abyss777) [inline web/app/view/ReportController.js]: I was thinking about it too. But may be do it for reports with one position only?

Comment (tananaev) [inline web/app/view/ReportController.js]: I think we can just select first item for any report. Even if it's now just one element. Kind of makes sense.

File: web/app/view/MapMarkerController.js (hunk 1 above)
Comment (tananaev) [inline web/app/view/MapMarkerController.js]: I mean why can't we just rely on `this.getView().getMapView().fit`?

**Comment by Abyss777** [comment]
Did the changes.

File: web/app/view/MapMarkerController.js (hunk 1 above)
Comment (Abyss777) [inline web/app/view/MapMarkerController.js]: Fit accept geometry or rectangle. We have a lot of geometries and have to find such rectangle that contain all geometries.

File: web/app/view/ReportController.js (hunk 2 above)
Comment (Abyss777) [inline web/app/view/ReportController.js]: If we select first position for every event report, then we have to select event in report grid for consistent.
Also if we select everything (in three panels) it should repeat behavior of manual selection and move map to selected position. It is not good, it should fit all positions after  generating report.

Comment (tananaev) [inline web/app/view/ReportController.js]: OK, then your option of auto-selecting if a single sounds better.

**Comment by tananaev** [comment]
Merged, thanks.

**Pull request merged**